src/main/java/com/startsteps/ecommerceapi/user/service/UserServiceImpl.java

```diff
@@ -1,42 +1,134 @@
+        }
+        User newUser = new User((user.getUserId()), user.getUsername(), passwordEncoder.encode(user.getPassword()),
+                user.getEmail(),false, true);
+        return userRepository.save(newUser);
+    }
+    @Override //TODO: authorized method
+    public User registerAdmin(UserDTO user) {
+        return null;
+    }
 
     @Override
-    public List<User> findAllUsers() {
-        return userRepository.findAll();
+    public User resetPasswordSendEmail(String userEmail) {
+        if(!userRepository.existsByEmail(userEmail) ) {
+            throw new UserNotFoundException("User with the email " + userEmail + " not found");
+        }
+        User user = userRepository.findByEmail(userEmail).orElseThrow(() ->
+                new UserNotFoundException("User with the email " + userEmail + " not found"));
+        String resetToken = generateResetToken();
+        createPasswordResetTokenForUser(user, resetToken);
+        SimpleMailMessage email = new SimpleMailMessage();
+        email.setTo(userEmail);
+        email.setSubject("Reset your password");
+        email.setText("To reset your password, click the link below:\n" +
+                "http://ecommerce-api/reset-password?token=" + resetToken);
+        return user;
     }
 
     @Override
-    public User addUser(User user) {
-        return userRepository.save(user);
+    public void createPasswordResetTokenForUser(final User user, final String token) {
+        final PasswordResetToken myToken = new PasswordResetToken(token, user);
+        passwordResetTokenRepository.save(myToken);
     }
 
     @Override
-    public void deleteUser(Long id) {
-        userRepository.deleteById(id);
+    public boolean changePassword(String token, String newPassword){
+       PasswordResetToken passwordResetToken = passwordResetTokenRepository.findByToken(token)
+               .orElseThrow(() -> new NoSuchElementException("Token not found: " + token));
+       Date tokenCreationDate = passwordResetToken.getCreatedDate();
+       if(isTokenExpired(tokenCreationDate)){
+           System.out.println("Token has been expired. Try again");
+           return false;
+       }
+       User user = passwordResetToken.getUser();
+       user.setPassword(passwordEncoder.encode(newPassword));
+       userRepository.save(user);
+       passwordResetTokenRepository.delete(passwordResetToken);
+       return true;
+   }
+    @Override
+    public Optional<PasswordResetToken> getPasswordResetToken(final String token) {
+        return passwordResetTokenRepository.findByToken(token);
+    }
+    public boolean isTokenExpired(final Date tokenCreationDate) {
+        LocalDateTime tokenCreationDateTime = tokenCreationDate.toInstant()
+                .atZone(ZoneId.systemDefault())
+                .toLocalDateTime();
+        LocalDateTime now = LocalDateTime.now();
+        Duration diff = Duration.between(tokenCreationDateTime, now);
+        return diff.toMinutes() >= 30;
```

Comment:
not seeing the JWT implementation right now but you can also get this from `Claims`, but also this might be an issue since you're using local time instead of GMT.

I register an account in Germany and I use this token on another device in Japan and my token is either worth a few more hours or already expired immediatelly.